### DISCUSSION TITLE
Managing async session nested transactions

### DISCUSSION THREAD

**Discussion by Nikolai-Dimitrov-Vidual** [sqlalchemy/sqlalchemy#11795] [Usage Questions] [answered]
Hey, so long story short I have a Pytest SQLalchemy (postgres+asyncpg) setup with a fixture that creates a connection, session, and transaction for every function tested, then closes the connection and session and rolls back the transaction at the end, like so:

```python
@pytest_asyncio.fixture(scope="function")
async def session(
    engine: AsyncEngine,
) -> AsyncGenerator[AsyncSession, None]:
    connection = await engine.connect()

    AsyncSessionMaker = sessionmaker(
        connection,
        class_=AsyncSession,
        expire_on_commit=False,
    )

    async_session = AsyncSessionMaker(bind=connection)

    main_transaction = await connection.begin()
    try:
        yield async_session
    finally:
        await async_session.close()
        await main_transaction.rollback()
        await connection.close()
```

This works, however it's very inefficient, since my (quite large) test setup with a lot of inserts and refreshes is stood up and torn down for every function. It would make a lot more sense for the session to be module scoped, and have a function scoped helper fixture that creates and rolls back transactions for each function. I read docs and questions and this was the best I could come up with:

```python
@pytest_asyncio.fixture(scope="module")
async def setup_session(
    engine: AsyncEngine,
) -> AsyncGenerator[AsyncSession, None]:
    connection = await engine.connect()

    AsyncSessionMaker = sessionmaker(
        connection,
        class_=AsyncSession,
        expire_on_commit=False,
    )

    async_session = AsyncSessionMaker(bind=connection)

    main_transaction = await connection.begin()
    try:
        yield async_session
    finally:
        await async_session.close()
        await main_transaction.rollback()
        await connection.close()


@pytest_asyncio.fixture(scope="function")
async def session(
    setup_session: AsyncSession,
) -> AsyncGenerator[AsyncSession, None]:
    transaction = await setup_session.begin_nested()
    try:
        yield setup_session
    finally:
        await transaction.rollback()
```

However, it doesn't work. Tests run ok individually, albeit with a `This transaction is closed` error every time, but when they're ran together it becomes apparent that the state is not reset after each function.

Here's what a typical test looks like:

```python
@pytest.mark.asyncio()
async def test_get_guides_succeeds(
    client: AsyncClient, session: AsyncSession
) -> None:
    app.dependency_overrides[get_current_user] = lambda: setup.user_empowered
    app.dependency_overrides[get_session] = lambda: session

    response = await client.get(
        f"/guides?guide_ids={setup.guide1.id}&guide_ids={setup.guide2.id}"
    )

    assert response.status_code == 200
    assert response.json()[0].get("id") == setup.guide1_return.id
    assert response.json()[1].get("id") == setup.guide2_return.id
```

With `Setup.create_test_setup(session)` being a bunch of adds, commits, and refreshes using the session to setup objects.
I also tried a bunch of other different configurations, but none of them worked. 
What am I doing wrong here? Any advice is very welcome.

**Comment by zzzeek** [comment]
I think the use of "module" scope for the connection + test transaction is going to produce problems and is likely not appropriate here.   we assume here you've ported the sync example from https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites into pytest + async, that's great.  But note the example is using setUp/tearDown which are test-scoped functions.  you want to run each individual test in a transaction, and fully roll back at the end.

the actual issue you have is that your INSERTs of fixture data take a long time.  I would do *those* at the module scope, then tear *those* down (using DELETE, DROP TABLE, etc. ) outside of the transactional portion of your test.

**Reply by Nikolai-Dimitrov-Vidual** [reply]
Yeah, the actual stand up of the setup would have been moved to a module scoped fixture instead of each test if I had managed to make it work, I've updated it to reflect that in the sample code.

What you're saying about just issuing SQL to revert the test setup makes sense. I thought it'd be clever to wrap it in a transaction that gets rolled back at the end of the module, and have each test be its own nested transaction, as it's much less code changes and with the risk of accidentally truncating the wrong database it seems a bit dangerous.

It feels like it should be very close to working, but I can't quite make it happen, so I implemented what I think you suggested and it works well for now.

Here's the code in case it helps anyone:

The relevant conftest.py fixtures (somewhat improved):

```python
@pytest_asyncio.fixture(scope="module")
async def AsyncSessionMaker(set_db_url) -> AsyncIterator[sessionmaker]:
    """
    Sets up the module scoped engine and connection and yields a sessionmaker to create sessions with.
    """
    engine = create_async_engine(DB_URL)  # , echo=True)

    if os.getenv("ENV_STATE") != "TEST":
        raise EnvironmentException(
            f'The ENV_STATE environment variable must be set to "{EnvStates.TEST}" for this operation.'
        )

    connection = await engine.connect()
    AsyncSessionMaker = sessionmaker(
        bind=connection,
        class_=AsyncSession,
        expire_on_commit=False,
    )
    try:
        yield AsyncSessionMaker
    finally:
        await engine.dispose()
        await connection.close()


@pytest_asyncio.fixture(scope="function")
async def session(
    AsyncSessionMaker: sessionmaker,
) -> AsyncIterator[AsyncSession]:
    """
    Yields a function scoped session.
    """

    async_session: AsyncSession = AsyncSessionMaker()

    try:
        yield async_session
    finally:
        await async_session.close()
```

The tear_down method (for the Setup class), fixture, and a sample test (the setup is now passed as a parameter):

```python
    @staticmethod
    async def tear_down(session: AsyncSession):
        if os.getenv("ENV_STATE") != EnvStates.TEST:
            raise EnvironmentException(
                f'The ENV_STATE environment variable must be set to "{EnvStates.TEST}" for this operation.'
            )
        statement = text(
                f"TRUNCATE {",".join(f'"{table.name}"' for table in reversed(Base.metadata.sorted_tables))} RESTART IDENTITY CASCADE;"
            )
        await session.execute(statement)
        await session.commit()


@pytest_asyncio.fixture(scope="module")
async def setup(AsyncSessionMaker: sessionmaker) -> AsyncGenerator[Setup, None]:
    async with AsyncSessionMaker() as session:
        setup = await Setup.create(session)
        try:
            yield setup
        finally:
            await Setup.tear_down(session)


@pytest.mark.asyncio()
async def test_get_guides_succeeds(
    client: AsyncClient, session: AsyncSession, setup: Setup
) -> None:
    app.dependency_overrides[get_current_user] = lambda: setup.user_empowered
    app.dependency_overrides[get_session] = lambda: session

    response = await client.get(
        f"/guides?guide_ids={setup.guide1.id}&guide_ids={setup.guide2.id}"
    )

    assert response.status_code == 200
    assert response.json()[1].get("id") == setup.guide1_return.id
    assert response.json()[0].get("id") == setup.guide2_return.id
```

Now, if you or anyone figures out how to do what I originally intended , which I would still prefer since it seems safer, do share.

**Reply by zzzeek** [reply]
doing INSERT/DELETE in a module level fixture and then doing the "rollback" thing in a test-level fixture is "safer" than trying to run lots of tests in one big transaction because the latter is going to explode badly if any one test has problems, and also having a transaction open means whatever effects have accumulated in that transaction are going to leak between tests.

as far as "the wrong database" being targeted, usually test suites run on a non-production database dedicated to tests and if it's a lightweight database like PostgreSQL, your test suite could spin up a local PostgreSQL in a container, which is a standard approach these days to database related tests.

**Reply by Nikolai-Dimitrov-Vidual** [reply]
No I got you, the only difference would be that the rollback would be a rollback of the transaction that created the setup, rather than truncating all the tables (and it'd be nice to touch only the main fixtures and each module, instead of that + all the tests). But fair points, and spinning up a new container for a DB to test on sounds pretty cool, certainly more portable than the manually created local server I have right now. Thanks!

**Reply by Nikolai-Dimitrov-Vidual** [reply, marked as answer]
Eventually I managed to get it working the way I wanted to. Here's the fixture setup:

```python
@pytest.fixture(scope="module")
async def engine(set_db_url) -> AsyncIterator[AsyncEngine]:
    engine = create_async_engine(DB_URL)  # , echo=True)
    if os.getenv("ENV_STATE") != "TEST":
        raise ApplicationException(
            'The ENV_STATE environment variable is not set to "TEST"'
        )
    yield engine
    await engine.dispose()


@pytest.fixture(scope="module")
async def connection(engine: AsyncEngine) -> AsyncIterator[AsyncConnection]:
    async with engine.connect() as connection:
        async with connection.begin() as transaction:
            yield connection
            await transaction.rollback()


@pytest.fixture(scope="module")
async def base_session(connection) -> AsyncIterator[AsyncSession]:
    async_session = AsyncSession(bind=connection, expire_on_commit=False)
    try:
        yield async_session
    finally:
        await async_session.close()


@pytest.fixture(scope="function")
async def session(connection) -> AsyncIterator[AsyncSession]:
    async with connection.begin_nested():
        async_session = AsyncSession(bind=connection, expire_on_commit=False)
        try:
            yield async_session
        finally:
            await async_session.close()
```

The base session fixture handles creating and automatically rolling back the test setup for each module, while the regular session fixture does the same for each test.

**Reply by zzzeek** [reply]
hmm what does the base_session fixture actually do?   I dont see it changing the state of anything and it's not called upon anywhere

**Reply by Nikolai-Dimitrov-Vidual** [reply]
This was for the conftest.py file, the `base_session` fixture is called and used once in every module of tests, it's what the test setup for the module is bound to, like so:

```python
@pytest.fixture(scope="module")
async def setup(base_session: AsyncSession) -> AsyncGenerator[Setup, None]:
    setup = await Setup.create(base_session)
    yield setup
```

Then in each test you'll use the regular `session`. This way only the changes made during the test get rolled back after each test and the setup for the module is only created and rolled back once. There's also no leakage between tests and runs.

**Reply by zzzeek** [reply]
oh that's your INSERT stuff, OK

**Reply by CaselIT** [reply]
If I understand correctly you are doing the following:
- start transaction
- insert data once
- for each test 
    - create savepoint
    - run test
    - release savepoint
- rollback transaction

**Reply by Nikolai-Dimitrov-Vidual** [reply]
Yeah pretty much.
